app/controllers/labels_controller.rb

```diff
@@ -3,7 +3,21 @@
 class LabelsController < ApplicationController
   load_and_authorize_resource
 
+  before_action :sanitize_page_param, only: :search
   def search
-    @labels = Label.order(:name)
+    results = Label.ransack(name_i_cont: params[:search]).result
+    paginated = results.paginate(per_page: 3, page: @page)
+
+    render json: {
+      results: paginated.map {|l| {id: l.name, text: l.name, label_color: l.color, label_font_color: l.font_color} },
+      pagination: {more: (@page * 3 < results.length)},
```

Comment:
`results.length` is not optimal here, it will render the pagination somewhat useless. In order to understand why, one has to look what happens here:

The `length` method will get the total number of entries of an array. At this point, `results` is not an array, but rather an `ActiveRecord_Relation`. This is a wrapper for the database call, and the actual results or only lazy loaded, when needed. Since the pagination only uses the database query and further customizes it (by limiting the result size and specifying an offset), this object (and thus all label objects) still aren't loaded. However, keep in mind that the original database query is unchanged (i.e., with the limit clause). Now, you are getting the length of the result array. Hence, the original query has to be fulfilled, in order to answer this question. Consequently, all matching labels will be fetched from the database, instantiated in an object and put into this array. Once everything is done, you get the length. As you see, we were still doing the regular processing for all matching labels, and the pagination didn't kick in. In this case, it would have been better to query the database just for the one number you're interested in: The result size. This would be done with a `COUNT(*)` query, and Active Record allows you to do so easily: `results.count`. This way, we omit transferring all labels and instantiating them, and leaving the hard work to the database.

Try the difference locally, you'll notice a change in the database query:

```
Label Load (2.8ms)  SELECT "labels".* FROM "labels"
Label Count (1.6ms)  SELECT COUNT(*) FROM "labels"
```

This change is better, but still not perfect. If the `paginate` already knows how many objects we have, we aren't using the result. Hence, we should leave it to paginate to decide whether another database query is needed for the overall number of results. Do so with `paginated.total_entries`. [Here](https://github.com/mislav/will_paginate/blob/8aa3a87022e4c17cb6ebf2ec0c9467bcdfd76b18/lib/will_paginate/active_record.rb#L68-L79) is the definition of the method. As you see, it will try calculating the overall size without further querying the database if possible, and otherwise switch to a database query using a raw `COUNT`. Therefore, we should use this method.